tests/functional/conftest.py

```diff
@@ -1,138 +1,158 @@
 from tests.scripts.pds_request import GenericPdsRequestor, PdsRecord
 import pytest
-import asyncio
-from api_test_utils.oauth_helper import OauthHelper
-from api_test_utils.apigee_api_apps import ApigeeApiDeveloperApps
-from api_test_utils.apigee_api_products import ApigeeApiProducts
+
+from .utils.apigee_api_apps import ApigeeApiDeveloperApps
+from .utils.apigee_api_products import ApigeeApiProducts
 from .config_files import config
 import random
-from time import time
 import json
 from typing import Union
 
-from .config_files.config import BASE_URL, CLIENT_ID, CLIENT_SECRET, REDIRECT_URI
 
+from pytest_nhsd_apim.identity_service import (
+    ClientCredentialsConfig,
+    ClientCredentialsAuthenticator,
+)
+
+from pytest_nhsd_apim.apigee_apis import (
+    ApigeeClient,
+    ApiProductsAPI,
+    ApigeeNonProdCredentials,
+    DeveloperAppsAPI,
+)
+
+import logging
+
+LOGGER = logging.getLogger(__name__)
+DEVELOPER_EMAIL = "apm-testing-internal-dev@nhs.net"
+
+
+@pytest.fixture()
+def client():
+    config = ApigeeNonProdCredentials()
+    return ApigeeClient(config=config)
+
+
+@pytest.fixture()
+def api_products(client):
+    return ApiProductsAPI(client=client)
+
+
+@pytest.fixture()
+def developer_apps(client):
+    return DeveloperAppsAPI(client=client)
 
-async def _set_default_rate_limit(product: ApigeeApiProducts):
+
+@pytest.fixture()
+def add_asid_to_testapp(developer_apps, nhsd_apim_test_app):
+    app = nhsd_apim_test_app()
+    LOGGER.info(f'app:{app}')
+    app_name = app["name"]
+
+    # Check if the ASID attribute is already available
+    app_attributes = developer_apps.get_app_attributes(email=DEVELOPER_EMAIL, app_name=app_name)
+    custom_attributes = app_attributes['attribute']
+    existing_asid_attribute = None
+    for attribute in custom_attributes:
+        if attribute['name'] == 'asid':
+            existing_asid_attribute = attribute['value']
+
+    if not existing_asid_attribute and config.ENV.contains("internal_dev_asid"):
+        LOGGER.info(f'ASID attribute not found. Adding {config.ENV["internal_dev_asid"]} to {app_name}')
+        # Add ASID to the test app - To be refactored when we move to .feature files TODO
+        custom_attributes.append({"name": "asid", "value": config.ENV["internal_dev_asid"]})
+        data = {"attribute": custom_attributes}
+        response = developer_apps.post_app_attributes(email=DEVELOPER_EMAIL, app_name=app_name, body=data)
+        LOGGER.info(f'Test app updated with ASID attribute: {response}')
```

Comment:
There is a similarly named `add_asd_to_test_app` in another conftest.py file. Is it worth moving this fixture up a level, to /tests/conftest.py, so that we a re not duplicating code?

Reply:
I think the user-restricted tests should be moved under the functional group and use the Feature files (I'll raise a ticket for it today); once moved over, we'll only be using the add_asid_to_testapp from tests/functional/conftest